mantidproject/mantid
# Speed up findingDetectors for IPeaks in PredictFractionalPeaks

PredictFractionalPeaks was using a new raytracer for every peak it created and therefore throwing away the caching information about the bounding boxs of detectors / panels causing a huge slow down. Resolved by exposing the findDetector(raytracer) overload to IPeaks and using single raytracer for all calls.
Work also include small refactor for const autos

Description of work.

**To test:**

* Pascals script at ISIS should be a good test for this:
```
LoadRaw(Filename='WISH00040151.raw', OutputWorkspace='40151', LoadLogFiles=False)
CropWorkspace(InputWorkspace='40151', OutputWorkspace='40151', XMin=6000, XMax=99000)
ConvertUnits(InputWorkspace='40151', OutputWorkspace='40151', Target='dSpacing')
FindSXPeaks(InputWorkspace='40151', PeakFindingStrategy='AllPeaks', AbsoluteBackground=10, ResolutionStrategy='AbsoluteResolution', XResolution=0.10000000000000001, PhiResolution=2, TwoThetaResolution=2, OutputWorkspace='40151_peaks')
FindUBUsingLatticeParameters(PeaksWorkspace='40151_peaks', a=10.01, b=8.9100000000000001, c=5.5099999999999998, alpha=90, beta=107.51000000000001, gamma=90, Tolerance=0.20000000000000001, Iterations=5)
PredictPeaks(InputWorkspace='40151_peaks', WavelengthMin=0.5, WavelengthMax=10, ReflectionCondition='C-face centred', OutputWorkspace='40151_predicted')
PredictFractionalPeaks(Peaks='40151_predicted', FracPeaks='40151_predicted_frac', HOffset='0.6,-0.6', LOffset='-0.08,0.08', IncludeAllPeaksInRange=True, Hmin=-2, Hmax=2, Kmin=-2, Kmax=2, Lmin=-2, Lmax=2)
```

* Run the script and ensure that `PredictFractionalPeaks` is fast (~10secs) previously on release build this was running at around 2mins so ensure there is a speed up in this alg. 


Fixes #22060

**Release Notes** 
This bug was not in the last release (introduced as a by-product of instrument changes this release) hence no need to add it to the release notes

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
